feat(loader): implement markdown parsing in MathpixPDFReader

eliasjudin opened the pull request:
Add functionality to properly handle PDF content:
- Add parse_markdown_text_to_tables method to separate tables and text
- Fix load_data implementation to properly process documents
- Fix lazy_load_data method
- Improve document metadata handling for tables and text sections

The loader now correctly processes PDFs through Mathpix API and converts content to proper Document objects.

## Description

- Added functionality to properly parse and process PDF content in MathpixPDFReader:
  - Implemented parse_markdown_text_to_tables method to correctly separate tables and text sections
  - Fixed load_data implementation to process documents with proper metadata
  - Added lazy_load_data method for memory-efficient document loading
  - Improved document metadata handling for both tables and text sections
- The loader now correctly processes PDFs through Mathpix API and converts content to proper Document objects

## Type of change

- [x] New features (non-breaking change).
- [x] Bug fix (non-breaking change).
- [ ] Breaking change (fix or feature that would cause existing functionality not to work as expected).

## Checklist

- [x] I have performed a self-review of my code.
- [x] I have added thorough tests if it is a core feature.
- [ ] There is a reference to the original bug report and related work.
- [x] I have commented on my code, particularly in hard-to-understand areas.
- [x] The feature is well documented.

eliasjudin committed:
✨ feat(loader): implement markdown parsing in MathpixPDFReader

Add functionality to properly handle PDF content:
- Add parse_markdown_text_to_tables method to separate tables and text
- Fix load_data implementation to properly process documents
- Fix lazy_load_data method
- Improve document metadata handling for tables and text sections

The loader now correctly processes PDFs through Mathpix API and converts content to proper Document objects.

taprosoft commented:
@eliasjudin please check the failed CI and comments.

taprosoft commented on libs/kotaemon/kotaemon/loaders/mathpix_loader.py:
```diff
@@ -139,26 +155,79 @@ def clean_pdf(self, contents: str) -> str:
-            content = kwargs["response_content"]
+        """Load data from file path."""
+        file_path = Path(file) if isinstance(file, str) else file
+        return super().load_data(file_path, extra_info, **load_kwargs)
```
Should we remove this?

taprosoft commented on libs/kotaemon/kotaemon/loaders/mathpix_loader.py:
```diff
@@ -171,8 +240,99 @@ def load_data(
+    ) -> Generator[Document, None, None]:
+        """Lazy load data from file path."""
+        file_path = Path(file) if isinstance(file, str) else file
+        return super().lazy_load_data(file_path, extra_info, **load_kwargs)
```
Consider removing this as well

eliasjudin committed:
fix(loader): remove super() calls blocking MathpixPDFReader implementation

Remove early returns using super() in load_data and lazy_load_data methods that were preventing the actual implementation from being executed. This fixes the "not implemented" error while maintaining the full PDF reader functionality.

eliasjudin reviewed (commented):
this comment was unrelated

eliasjudin commented:
need to resolve this issue
```python
1 validation error for CiteEvidence
evidences
  Input should be a valid list [type=list_type, input_value='["Let $M$ be a multiplic...\in M$ and $x \\in X$"]', input_type=str]
    For further information visit https://errors.pydantic.dev/2.9/v/list_type
Got 0 cited docs
```
any suggestions?

eliasjudin commented:
page numbering is incorrect, all items say from page 1

eliasjudin commented:
> need to resolve this issue
> 
> ```python
> 1 validation error for CiteEvidence
> evidences
>   Input should be a valid list [type=list_type, input_value='["Let $M$ be a multiplic...\in M$ and $x \\in X$"]', input_type=str]
>     For further information visit https://errors.pydantic.dev/2.9/v/list_type
> Got 0 cited docs
> ```
> 
> any suggestions?

fixed this by adding check if str in `citation.py` and then wrapping to list
